lib/mongo/srv/resolver.rb

```diff
@@ -0,0 +1,65 @@
+# Copyright (C) 2014-2019 MongoDB, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the 'License');
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'resolv'
+
+module Mongo
+  module SRV
+
+    # The SRV::Resolver class encapsulates the necessary behavior for querying SRV records.
+    #
+    # @api private
+    class Resolver
+      include Loggable
+
+      RECORD_PREFIX = '_mongodb._tcp.'.freeze
+
+      def initialize(raise_on_invalid = false)
```

Comment:
Why is this an option, given that clients can rescue exceptions?

Reply:
The polling SRV records spec indicates that warnings should only be logged rather than errors raised. If we only catch the error that's raised externally, then an error with one record will prevent the driver from checking the rest of the records, which might be fine.

Reply:
Then why is there an option to raise exceptions? It also has no test coverage.

Reply:
The initial seedlist discovery spec requires that an error is raised if any invalid records are found, whereas the polling SRV records for mongos spec requires that errors are not raised. We need to be able to do different behavior for each of them, but most of the code is the same, which is why this is being factored out in the first place. The only reason I separated out this code is that you asked me to make a separate PR, but it seems like it's creating more work than it's saving.